flink-end-to-end-tests/flink-datastream-allround-test/src/main/java/org/apache/flink/streaming/tests/SlidingWindowCheckMapper.java

```diff
@@ -0,0 +1,178 @@
+import java.util.Optional;
+
+/**
+ * This mapper validates sliding event time window. It checks each event belongs to appropriate number of consecutive
+ * windows.
+ */
+public class SlidingWindowCheckMapper extends RichFlatMapFunction<Tuple2<Integer, List<Event>>, String> {
+
+	private static final long serialVersionUID = -744070793650644485L;
+
+	/** This value state tracks previously seen events with the number of windows they appeared in. */
+	private transient ValueState<List<Tuple2<Event, Integer>>> eventsSeenSoFar;
+
+	private transient ValueState<Long> lastSequenceNumber;
+
+	private final int slideFactor;
+
+	SlidingWindowCheckMapper(int slideFactor) {
+		this.slideFactor = slideFactor;
+	}
+
+	@Override
+	public void open(Configuration parameters) {
+		ValueStateDescriptor<List<Tuple2<Event, Integer>>> previousWindowDescriptor =
+			new ValueStateDescriptor<>("eventsSeenSoFar",
+				new ListTypeInfo<>(new TupleTypeInfo<>(TypeInformation.of(Event.class), BasicTypeInfo.INT_TYPE_INFO)));
+
+		eventsSeenSoFar = getRuntimeContext().getState(previousWindowDescriptor);
+
+		ValueStateDescriptor<Long> lastSequenceNumberDescriptor =
+			new ValueStateDescriptor<>("lastSequenceNumber", BasicTypeInfo.LONG_TYPE_INFO);
+
+		lastSequenceNumber = getRuntimeContext().getState(lastSequenceNumberDescriptor);
+	}
+
+	@Override
+	public void flatMap(Tuple2<Integer, List<Event>> value, Collector<String> out) throws Exception {
+		List<Tuple2<Event, Integer>> previousWindowValues = Optional.ofNullable(eventsSeenSoFar.value()).orElseGet(
+			Collections::emptyList);
+
+		List<Event> newValues = value.f1;
+		Optional<Event> lastEventInWindow = verifyWindowContiguity(newValues, out);
+
+		Long lastSequenceNumberSeenSoFar = lastSequenceNumber.value();
+		List<Tuple2<Event, Integer>> newWindows =
+			verifyPreviousOccurences(previousWindowValues, newValues, out, lastSequenceNumberSeenSoFar);
+
+		if (lastEventInWindow.isPresent()) {
+			updateLastSeenSequenceNumber(lastEventInWindow.get(), lastSequenceNumberSeenSoFar);
+		}
+
+		eventsSeenSoFar.update(newWindows);
+	}
+
+	private void updateLastSeenSequenceNumber(
+			Event lastEventInWindow,
+			Long lastSequenceNumberSeenSoFar) throws IOException {
+		long lastSequenceNumberInWindow = lastEventInWindow.getSequenceNumber();
+		if (lastSequenceNumberSeenSoFar == null || lastSequenceNumberInWindow > lastSequenceNumberSeenSoFar) {
+			lastSequenceNumber.update(lastSequenceNumberInWindow);
+		}
+	}
+
+	/**
+	 * Verifies if all values from previous windows appear in the new one. Returns union of all events seen so far that
+	 * were not seen <b>slideFactor</b> number of times yet.
+	 */
+	private List<Tuple2<Event, Integer>> verifyPreviousOccurences(
+			List<Tuple2<Event, Integer>> previousWindowValues,
+			List<Event> newValues,
+			Collector<String> out, Long lastSequenceNumberSeenSoFar) {
+		List<Tuple2<Event, Integer>> newEventsSeenSoFar = new ArrayList<>();
+		List<Event> seenWindows = new ArrayList<>();
```

Comment:
The `seenWindows` contain the seen events right? If yes, this could be renamed to `seenEvents`.